java/code/src/com/suse/manager/reactor/messaging/RegisterMinionEventMessageAction.java

```diff
@@ -127,69 +128,67 @@ public void execute(EventMessage msg) {
-                    }
-                    else if (ak.getServer() == null) {
-                        LOG.error("Management Key is not a reactivation key: " + managmentKeyLabel.get());
-                    }
-                    else {
-                        isReactivation = true;
-                    }
+    private void registerMinion(String minionId, boolean isSaltSSH, Optional<Long> saltSSHProxyId,
+                                Optional<String> actKeyOverride, String machineId) {
+        if (!duplicateMinionNamePresent(minionId)) {
+            boolean isReactivation = false;
+            ActivationKey ak = null;
+            ValueMap grains = new ValueMap(SALT_SERVICE.getGrains(minionId).orElseGet(HashMap::new));
+            Optional<String> managmentKeyLabel = getManagementKeyLabelFromGrains(grains);
+            if (managmentKeyLabel.isPresent()) {
+                ak = ActivationKeyFactory.lookupByKey(managmentKeyLabel.get());
+                if (ak != null && ak.getKickstartSession() != null) {
+                    ak.getKickstartSession().markComplete("Installation completed.");
                 }
-
-                Optional<User> creator = MinionPendingRegistrationService.getCreator(minionId);
-                if (!isReactivation &&
-                        checkIfMinionAlreadyRegistered(minionId, machineId, creator, isSaltSSH, grains)) {
-                    return true;
+                if (ak == null) {
+                    LOG.info("Outdated Management Key defined for " + minionId + ": " + managmentKeyLabel.get());
                 }
-                // Check if this minion id already exists
-                if (!isReactivation && duplicateMinionNamePresent(minionId)) {
-                    return false;
+                else if (ak.getServer() == null) {
+                    LOG.error("Management Key is not a reactivation key: " + managmentKeyLabel.get());
                 }
-                if (isReactivation) {
-                    reactivateSystem(minionId, machineId, creator, grains, Optional.ofNullable(ak));
+                else {
+                    isReactivation = true;
                 }
-                finalizeMinionRegistration(minionId, machineId, creator, saltSSHProxyId,
-                        actKeyOverride.isPresent() ? actKeyOverride : Optional.empty(),
-                        isSaltSSH, grains);
-                return true;
             }
-        );
+            Optional<User> creator = MinionPendingRegistrationService.getCreator(minionId);
+            if (!isReactivation &&
+                    checkIfMinionAlreadyRegistered(minionId, machineId, creator, isSaltSSH, grains)) {
+                return;
+            }
+            if (isReactivation) {
+                reactivateSystem(minionId, machineId, creator, grains, Optional.ofNullable(ak));
+            }
+            finalizeMinionRegistration(minionId, machineId, creator, saltSSHProxyId, actKeyOverride, isSaltSSH,
```

Comment:
I would really put grains required to detect reactivation 'susemanager:management_key' as startup grains as well. So we can ask for the rest of grains either here or directly in `finalizeMinionRegistration`.
We will then skip the grains call for all already registered minions only having their salt minions restarted.
Could that work?

Reply:
Ok thought about that but then reverted it back. The reason behind was that `management_key` is only needed for the new systems for 1st time registration and in that case, we needed a lot of grains and we are fine with making one salt call which returns all the grains. Of course, we can add all the grains in the startup event which are needed for the  registration but that involved a lot of grains.

 We can refactor but that should probably be in a separate PR and will need a lot more time and testing, it was bit risky. Here in this PR, main idea was to avoid any salt calls for the already registered systems and for that, we only needed `machine_id` and `saltboot_initrd` grains.